aptos-move/aptos-vm/src/sharded_block_executor/sharded_aggregator_service.rs

```diff
@@ -0,0 +1,308 @@
+        if self.is_positive == rhs.is_positive {
+            return Self {
+                delta: self.delta + rhs.delta,
+                is_positive: self.is_positive,
+            };
+        }
+
+        // the deltas are of opposite signs, we subtract the smaller from the larger and keep the
+        // sign of the larger
+        let (pos, neg) = if self.is_positive {
+            (self.delta, rhs.delta)
+        } else {
+            (rhs.delta, self.delta)
+        };
+
+        if pos >= neg {
+            return Self {
+                delta: pos - neg,
+                is_positive: true,
+            };
+        }
+        Self {
+            delta: neg - pos,
+            is_positive: false,
+        }
+    }
+}
+
+#[test]
+fn test_delta_u128() {
+    assert_eq!(DeltaU128::default().delta, 0);
+    assert!(DeltaU128::default().is_positive);
+
+    {
+        // get positive delta
+        let delta_pos = DeltaU128::get_delta(100, 50);
+        assert_eq!(delta_pos.delta, 50);
+        assert!(delta_pos.is_positive);
+
+        // get negative delta
+        let delta_neg = DeltaU128::get_delta(50, 100);
+        assert_eq!(delta_neg.delta, 50);
+        assert!(!delta_neg.is_positive);
+    }
+
+    {
+        // test add_delta
+        let delta1 = DeltaU128 {
+            delta: 100,
+            is_positive: true,
+        };
+        assert_eq!(delta1.add_delta(50), 150);
+
+        let delta2 = DeltaU128 {
+            delta: 50,
+            is_positive: false,
+        };
+        assert_eq!(delta2.add_delta(50), 0);
+    }
+
+    {
+        // test all cases for ops::Add
+        let delta1 = DeltaU128 {
+            delta: 100,
+            is_positive: true,
+        };
+        let delta2 = DeltaU128 {
+            delta: 50,
+            is_positive: false,
+        };
+        let delta3 = DeltaU128 {
+            delta: 100,
+            is_positive: true,
+        };
+        // checks for [pos > neg]; [pos, pos]
+        let mut delta_sum = delta1 + delta2 + delta3;
+        assert_eq!(delta_sum.delta, 150);
+        assert!(delta_sum.is_positive);
+
+        let delta4 = DeltaU128 {
+            delta: 500,
+            is_positive: false,
+        };
+        let delta5 = DeltaU128 {
+            delta: 200,
+            is_positive: false,
+        };
+        // checks for [neg > pos]; [neg, neg]
+        delta_sum = delta_sum + delta4 + delta5;
+        assert_eq!(delta_sum.delta, 550);
+        assert!(!delta_sum.is_positive);
+    }
+}
+
+fn is_total_supply_absent(
+    total_supply: Option<u128>,
+    shard_id: usize,
+    num_shards: usize,
+    txn_idx: usize,
+    txns_len: usize,
+) -> bool {
+    // only first and last txn of the block can have 'none' for 'total_supply'
+    if total_supply.is_none()
+        && ((shard_id == 0 && txn_idx == 0) || (shard_id == num_shards && txn_idx == txns_len))
+    {
+        return true;
+    }
+    false
+}
+
+fn is_total_supply_absent_global_shard(
```

Comment:
I am wondering instead of doing this, it will be cleaner if we just remove the first and last `TransactionOutput` from the `sharded_output`/ `global_output` and process things in parallel without having to worry about the total supply being present?

Reply:
You can probably do this with skip_first/last on the iterator pretty easily..

Reply:
As discussed offline we don't have to assert in sharded execution module that all txns except first and last need to have total_supply